proposals/function-references/Overview.md

```diff
@@ -0,0 +1,507 @@
+# Typed Function References for WebAssembly
+
+## Introduction
+
+This proposal adds function references that are typed and can be called directly. Unlike `funcref` and the existing `call_indirect` instruction, typed function references need not be stored into a table to be called (though they can). A typed function reference can be formed from any function index.
+
+The proposal distinguished regular and nullable function reference. The former cannot be null, and a call through them does not require any runtime check.
+
+The proposal has instructions for producing and consuming (calling) function references. It also includes instruction for testing and converting between regular and nullable references.
+
+Typed references have no canonical default value, because they cannot be null. To enable storing them in locals, which so far depend on default values for initialisation, the proposal also introduces a new instruction `let` for block-scoped locals whose initialisation values are taken from the operand stack.
+
+In addition to the above, we could also decide to include an instruction for forming a *closure* from a function reference, which takes a prefix of the function's arguments and returns a new function reference with those parameters bound. (Hence, conceptually, all function references would be closures of 0 or more parameters.)
+
+Note: In a Wasm engine, function references (whether first-class or as table entries) are already a form of closure since they must close over a specific module instance (its globals, tables, memory, etc) while their code is shared across multiple instances of the same module. It is hence expected that the ability to form language-level closures is not an observable extra cost.
+
+
+### Motivation
+
+* Enable efficient indirect function calls without runtime checks
+
+* Represent first-class function pointers without the need for tables
+
+* Easier and more efficient exchange of function references between modules and with host environment
+
+* Optionally, support for safe closures
+
+* Separate independently useful features from [GC proposal](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md)
+
+
+### Summary
+
+* This proposal is based on the [reference types proposal](https://github.com/WebAssembly/reference-types))
+
+* Add a new form of *typed reference type* `ref $t` and a nullable variant `(ref null $t)`, where `$t` is a type index; can be used as both a value type or an element type for tables
+
+* Add an instruction `ref.as_non_null` that converts a nullable reference to a non-nullable one or traps if null
+
+* Add an instruction `br_on_null` that converts a nullable reference to a non-nullable one or branches if null, as well as the inverted `br_on_non_null`
+
+* Add an instruction `call_ref` for calling a function through a `ref $t`
+
+* Refine the instruction `ref.func $f` to return a typed function reference
+
+* Optionally add an instruction `func.bind` to create a closure
```

Comment:
I was under the impression that func.bind was not happening.

Reply:
Yes, but I yet have to update the overview.